sbi/src/pmu.rs

```diff
@@ -0,0 +1,823 @@
+    pub fn get_minh(&self) -> bool {
+        self.0 & (1 << 7) != 0
+    }
+}
+
+/// This encapsulates the bit-fields for PMU start_flags parameter as described in the SBI documentation
+/// for sbi_pmu_counter_start
+#[derive(Copy, Clone, Debug, Default)]
+pub struct PmuCounterStartFlags(u64);
+
+impl PmuCounterStartFlags {
+    /// Constructs a new PmuCounterStartFlags from a valid passed-in value.
+    pub fn from_raw_value(value: u64) -> Result<Self> {
+        match value {
+            0 | 1 => Ok(PmuCounterStartFlags(value)),
+            _ => Err(Error::InvalidParam),
+        }
+    }
+
+    /// Returns the raw value of the inner bit-flag field.
+    pub fn raw(&self) -> u64 {
+        self.0
+    }
+
+    /// Sets the set_init_value bit-flag (set initial counter value).
+    pub fn set_init_value(self) -> Self {
+        PmuCounterStartFlags(self.0 | 1)
+    }
+
+    /// Returns if set_init_value bit-flag is set.
+    pub fn get_init_value(&self) -> bool {
+        self.0 & 1 != 0
+    }
+}
+
+/// This encapsulates the bit-fields for PMU stop_flags parameter as described in the SBI documentation
+/// for sbi_pmu_counter_stop
+#[derive(Copy, Clone, Debug, Default)]
+pub struct PmuCounterStopFlags(u64);
+
+impl PmuCounterStopFlags {
+    /// Constructs a new PmuCounterStopFlags from a valid passed-in value.
+    pub fn from_raw_value(value: u64) -> Result<Self> {
+        match value {
+            0 | 1 => Ok(PmuCounterStopFlags(value)),
+            _ => Err(Error::InvalidParam),
+        }
+    }
+
+    /// Returns the raw value of the inner bit-flag field.
+    pub fn raw(&self) -> u64 {
+        self.0
+    }
+
+    /// Sets the stop_reset bit-flag (resets the counter after stopping).
+    pub fn set_reset_flag(self) -> Self {
+        PmuCounterStopFlags(self.0 | 1)
+    }
+
+    /// Returns if the stop_reset bit flag is set.
+    pub fn get_reset_flag(&self) -> bool {
+        self.0 & 1 != 0
+    }
+}
+
+/// This encapsulates the counter information returned by the call to sbi_pmu_counter_get_info.
+#[derive(Copy, Clone, Debug)]
+pub struct PmuCounterInfo(u64);
+
+impl PmuCounterInfo {
+    /// Constructs a PmuCounterInfo from the passed in value.
+    pub fn new(value: u64) -> Self {
+        PmuCounterInfo(value)
+    }
+
+    /// Returns the inner value.
+    pub fn raw(&self) -> u64 {
+        self.0
+    }
+
+    /// Returns if the counter is a hardware counter that uses the supported CSR range.
+    pub fn is_hardware_counter(&self) -> bool {
+        const CSR_CYCLE: u64 = 0xC00;
+        const CSR_HPMCOUNTER31: u64 = 0xC1F;
+        self.0 & (1 << 63) == 0 && self.get_csr() >= CSR_CYCLE && self.get_csr() <= CSR_HPMCOUNTER31
+    }
+
+    /// Returns if the counter is a firmware counter.
+    pub fn is_firmware_counter(&self) -> bool {
+        self.0 & (1 << 63) != 0
+    }
+
+    /// Returns the 12-bit CSR number associated with the counter.
+    pub fn get_csr(&self) -> u64 {
+        self.0 & 0b1111_1111_1111_u64
```

Comment:
Is this a standard rust coding style ? It is painful for a C developer :)

Reply:
I don't have a strong opinion here, but I personally find hex masks easier to read when dealing with wide, contiguous bitifelds like this.